source/common/http/http2/codec_impl.cc

```diff
@@ -544,6 +543,31 @@ void ConnectionImpl::ClientStreamImpl::decodeHeaders() {
   } else {
     response_decoder_.decodeHeaders(std::move(headers), sendEndStream());
   }
+#else
+  // in UHV mode the :status header at this point can be malformed, as it is validated
+  // later on in the response_decoder_.decodeHeaders() call.
+  // Account for this here.
+  absl::optional<uint64_t> status_opt = Http::Utility::getResponseStatusOrNullopt(*headers);
+  if (status_opt.has_value()) {
+    const uint64_t status = status_opt.value();
+    // Non-informational headers are non-1xx OR 101-SwitchingProtocols, since 101 implies that
+    // further proxying is on an upgrade path.
+    received_noninformational_headers_ =
+        !CodeUtility::is1xx(status) || status == enumToInt(Http::Code::SwitchingProtocols);
```

Comment:
Good point. Both balsa and EnvoyQuicClientStream validate status, so it does not reach UHV. I will work on fixing this. Balsa fix will need to wait, since Bence is OOO.

Reply:
I have moved the `:status` header checks for H/2 and H/3 codecs into UHV when it is enabled. For H/1 codec it is a lot more difficult to move status validation in the response line out of the low level parser (http-parser and Balsa), as it will require changing parser API and a few other spots where parser itself is using the code. This may become more tractable once http-parser is decommissioned.

I think keeping status validation in H/1 parser for now is ok, as it just reduces visibility into protocol errors if status is malformed. When the plumbing for the parser error codes into reset reason strings is finished it will reduce this shortcoming.